lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/authentication/KerberosConfiguration.java

```diff
@@ -0,0 +1,146 @@
+        {
+            @Override
+            public AppConfigurationEntry[] getAppConfigurationEntry(String name)
+            {
+                return new AppConfigurationEntry[] {
+                        new AppConfigurationEntry(
+                                KERBEROS_LOGIN_MODULE,
+                                AppConfigurationEntry.LoginModuleControlFlag.REQUIRED,
+                                options)};
+            }
+        };
+    }
+
+    public static class Builder
+    {
+        private String kerberosPrincipal;
+        private Optional<String> keytabLocation = Optional.empty();
+        private Optional<String> credentialCacheLocation = Optional.empty();
+
+        public Builder forKerberosPrincipal(String kerberosPrincipal)
+        {
+            this.kerberosPrincipal = kerberosPrincipal;
+            return this;
+        }
+
+        public Builder withkeytabLocation(String keytabLocation)
+        {
+            verifyFile(keytabLocation);
+            this.keytabLocation = Optional.of(keytabLocation);
+            return this;
+        }
+
+        public Builder withCredentialCacheLocation(String credentialCacheLocation)
+        {
+            verifyFile(credentialCacheLocation);
+            this.credentialCacheLocation = Optional.of(credentialCacheLocation);
+            return this;
+        }
+
+        public KerberosConfiguration build()
+        {
+            ImmutableMap.Builder<String, String> optionsBuilder = ImmutableMap.<String, String>builder()
+                    .put("doNotPrompt", "true")
+                    .put("isInitiator", "true")
+                    .put("principal", kerberosPrincipal);
+
+            keytabLocation.ifPresent(
+                    keytab -> optionsBuilder
+                            .put("storeKey", "true")
+                            .put("useKeyTab", "true")
+                            .put("keyTab", keytab));
+
+            credentialCacheLocation.ifPresent(
+                    credentialCache -> optionsBuilder
+                            .put("useTicketCache", "true")
+                            .put("renewTGT", "true")
+                            .put("ticketCache", credentialCache));
+
+            return new KerberosConfiguration(createKerberosPrincipal(kerberosPrincipal), optionsBuilder.buildOrThrow());
+        }
+
+        private static KerberosPrincipal createKerberosPrincipal(String principal)
+        {
+            try {
+                return new KerberosPrincipal(getServerPrincipal(principal, InetAddress.getLocalHost().getCanonicalHostName()));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+
+        private static String getServerPrincipal(String principal, String hostname)
+        {
+            String[] components = principal.split("[/@]");
+            if (components.length != 3 || !components[1].equals(HOSTNAME_PATTERN)) {
```

Comment:
I don't know whether it's a correct implementation.

Principal could be i.e. `hive/_HOST`

Can we just `principal.replaceFirst(HOSTNAME_PATTERN, hostname.toLowerCase(ENGLISH))`

Reply:
Realm can be optional